### ISSUE TITLE
YoloV11 CPU Memory Leaks When Access Inference Results

### ISSUE THREAD

**Issue by alien087** [ultralytics/ultralytics#21241]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

![Image](https://github.com/user-attachments/assets/6d0299fc-f245-459c-8ff0-32fdcb078802)

Hi, I have a code like this. but I have a problem when doing inference. So when I access the variable "results" via for or (next), the memory usage increases over time. But if I delete the code lines 200-201, there is no increase in memory usage. 

Can anyone help me? because if I run this for 1 hour, it can eat up gigabytes of RAM

### Environment

Ultralytics 8.3.105 🚀 Python-3.10.18 torch-2.6.0+cpu CPU (AMD Ryzen 7 8845HS w/ Radeon 780M Graphics)
Setup complete ✅ (16 CPUs, 15.3 GB RAM, 371.7/475.9 GB disk)

OS                  Windows-10-10.0.26100-SP0
Environment         Windows
Python              3.10.18
Install             pip
Path                C:\Users\dspit\anaconda3\envs\map_test2\Lib\site-packages\ultralytics
RAM                 15.26 GB
Disk                371.7/475.9 GB
CPU                 AMD Ryzen 7 8845HS w/ Radeon 780M Graphics
CPU count           16
GPU                 None
GPU count           None
CUDA                None

numpy               ✅ 1.26.4<=2.1.1,>=1.23.0
matplotlib          ✅ 3.10.3>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.2.1>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.15.2>=1.4.1
torch               ✅ 2.6.0>=1.8.0
torch               ✅ 2.6.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.21.0>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0

### Minimal Reproducible Example

```
# How I Import model
    det_model_path_str = os.path.join(base_dir, "models", "yolo11n_openvino_model", "yolo11n.xml")
    int8_model_det_path = os.path.join(base_dir, "models", "int8", "yolo11n_openvino_int8_model", "yolo11n.xml")

    # Initialize OpenVINO
    core = ov.Core()
    ov_config = {}
    det_ov_model = core.read_model(int8_model_det_path)
    det_compiled_model = core.compile_model(det_ov_model, "CPU", ov_config)

    # Prepare YOLO model
    det_model_dir = os.path.dirname(det_model_path_str)
    det_model = YOLO(det_model_dir, task="detect")

    if det_model.predictor is None:
        custom = {"conf": 0.25, "batch": 1, "save": False, "mode": "predict"}
        args = {**det_model.overrides, **custom}
        det_model.predictor = det_model._smart_load("predictor")(overrides=args, _callbacks=det_model.callbacks)
        det_model.predictor.setup_model(model=det_model.model)

    det_model.predictor.model.ov_compiled_model = det_compiled_model
```

```
#How I Inference
results = model(frame,  stream=True, iou=0.45, conf=0.2, classes=[0], verbose=False)
    print(f"Ukuran results: {sys.getsizeof(results)} bytes")

    for r in results:
       pass
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @alien087, thank you for reporting this and for providing detailed environment info and a code snippet! 🚀 This is an automated response to help you get the fastest support. An Ultralytics engineer will also review and assist you soon.

If this is a 🐛 Bug Report, please ensure you have provided a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/). If possible, try to further narrow down your code to the smallest example that reproduces the memory leak behavior. This will help us debug and resolve the issue much faster.

For reference and further troubleshooting, we recommend:
- Checking the [Docs](https://docs.ultralytics.com/) for [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples.
- Reviewing our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) if your issue is related to custom training.

Join the Ultralytics community for support and discussion:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time chat
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://www.reddit.com/r/Ultralytics/) for sharing knowledge and questions

## Upgrade

Please upgrade to the latest `ultralytics` package and verify all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to check if the issue persists in the most recent version:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your contribution!

**Comment by Y-T-G** [comment]
> But if I delete the code lines 200-201, there is no increase in memory usage.

If you delete those lines, then there would be no inference 

> Can anyone help me? because if I run this for 1 hour, it can eat up gigabytes of RAM

It should automatically release old memory unless you're storing the results in memory

**Comment by alien087** [comment]
> > But if I delete the code lines 200-201, there is no increase in memory usage.
> 
> If you delete those lines, then there would be no inference
> 
> > Can anyone help me? because if I run this for 1 hour, it can eat up gigabytes of RAM
> 
> It should automatically release old memory unless you're storing the results in memory

so, what's should i do? i tried to del veriable results after "for" but the result is the same. Can you help me, please?

**Comment by Y-T-G** [comment]
You don't have to worry about it because it will release the memory automatically

**Comment by alien087** [comment]
> You don't have to worry about it because it will release the memory automatically

no, it;'s not. in my case, the memory will growth until 10GB after several hours

I Already tried with another backend (pytorch and onnx) and memory is fine. but when i'm using openvino, the memory will growth

**Comment by Y-T-G** [comment]
In your code, you're making some changes manually like creating the predictor. You can try without making such changes. And also try upgrading to the latest Ultralytics.

**Comment by Y-T-G** [comment]
Ultralytics lets you select the device for OpenVINO in the latest version, so you don't need to manually do that in the code.

https://docs.ultralytics.com/integrations/openvino/#__tabbed_1_1

**Comment by alien087** [comment]
> In your code, you're making some changes manually like creating the predictor. You can try without making such changes. And also try upgrading to the latest Ultralytics.

Hey Thank you!!! Appreantly it's because the custom predictor that i've made. when i using default configuration the memory is stable. Thanks a Lot!!

**Comment by glenn-jocher** [comment]
Great to hear you resolved the memory leak issue! As you discovered, manually creating and configuring the predictor can sometimes interfere with Ultralytics' internal memory management. The default configuration handles memory cleanup automatically, which is why switching back to it fixed the problem. This is a good reminder that the standard inference approach is usually the most reliable for memory management.

**Comment by erfan-zekri** [comment]
Hi @glenn-jocher and @Y-T-G 

I have the same problem
I run the following simple code on both my amd64 laptop and arm64 raspberry pi 5
I didn't see any memory leak on my local laptop but I experienced gradual memory usage on raspberry pi 5
Do you have any solution to solve this memory leak problem?

code:
```
from ultralytics import YOLO

model = YOLO(
            "models/yolo11n_openvino_model/"
        )

while True:
    model()
```

### Raspberry pi 5:

Ultralytics 8.3.162 🚀 Python-3.11.2 torch-2.7.1+cpu CPU (Cortex-A76)
Setup complete ✅ (4 CPUs, 7.9 GB RAM, 29.9/58.0 GB disk)

OS                  Linux-6.12.25+rpt-rpi-2712-aarch64-with-glibc2.36
Environment         Linux
Python              3.11.2
Install             git
Path                /home/pi/zekri/projects/gitlab/vms/pi/anpr/.venv/lib/python3.11/site-packages/ultralytics
RAM                 7.87 GB
Disk                29.9/58.0 GB
CPU                 Cortex-A76
CPU count           4
GPU                 None
GPU count           None
CUDA                None

numpy               ✅ 2.2.6>=1.23.0
matplotlib          ✅ 3.10.3>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.2.1>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.4>=2.23.0
scipy               ✅ 1.16.0>=1.4.1
torch               ✅ 2.7.1>=1.8.0
torch               ✅ 2.7.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.22.1>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.3.0>=1.1.4
ultralytics-thop    ✅ 2.0.14>=2.0.0


### Laptop:

Ultralytics 8.3.162 🚀 Python-3.12.11 torch-2.7.1+cu126 CUDA:0 (NVIDIA GeForce GTX 850M, 4038MiB)
Setup complete ✅ (8 CPUs, 15.5 GB RAM, 435.6/455.0 GB disk)

OS                  Linux-6.8.0-60-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.12.11
Install             git
Path                /home/developer/zekri/projects/gitlab/vms/raspberry/ai/anpr/.venv/lib/python3.12/site-packages/ultralytics
RAM                 15.51 GB
Disk                435.6/455.0 GB
CPU                 Intel Core(TM) i7-4710HQ 2.50GHz
CPU count           8
GPU                 NVIDIA GeForce GTX 850M, 4038MiB
GPU count           1
CUDA                12.6

numpy               ✅ 2.2.6>=1.23.0
matplotlib          ✅ 3.10.3>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.2.1>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.4>=2.23.0
scipy               ✅ 1.16.0>=1.4.1
torch               ✅ 2.7.1>=1.8.0
torch               ✅ 2.7.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.22.1>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.3.0>=1.1.4
ultralytics-thop    ✅ 2.0.14>=2.0.0

**Comment by alien087** [comment]
> Hi @glenn-jocher and @Y-T-G 
> 
> I have the same problem
> I run the following simple code on both my amd64 laptop and arm64 raspberry pi 5
> I didn't see any memory leak on my local laptop but I experienced gradual memory usage on raspberry pi 5
> Do you have any solution to solve this memory leak problem?
> 
> code:
> ```
> from ultralytics import YOLO
> 
> model = YOLO(
>             "models/yolo11n_openvino_model/"
>         )
> 
> while True:
>     model()
> ```
> 
> ### Raspberry pi 5:
> 
> Ultralytics 8.3.162 🚀 Python-3.11.2 torch-2.7.1+cpu CPU (Cortex-A76)
> Setup complete ✅ (4 CPUs, 7.9 GB RAM, 29.9/58.0 GB disk)
> 
> OS                  Linux-6.12.25+rpt-rpi-2712-aarch64-with-glibc2.36
> Environment         Linux
> Python              3.11.2
> Install             git
> Path                /home/pi/zekri/projects/gitlab/vms/pi/anpr/.venv/lib/python3.11/site-packages/ultralytics
> RAM                 7.87 GB
> Disk                29.9/58.0 GB
> CPU                 Cortex-A76
> CPU count           4
> GPU                 None
> GPU count           None
> CUDA                None
> 
> numpy               ✅ 2.2.6>=1.23.0
> matplotlib          ✅ 3.10.3>=3.3.0
> opencv-python       ✅ 4.11.0.86>=4.6.0
> pillow              ✅ 11.2.1>=7.1.2
> pyyaml              ✅ 6.0.2>=5.3.1
> requests            ✅ 2.32.4>=2.23.0
> scipy               ✅ 1.16.0>=1.4.1
> torch               ✅ 2.7.1>=1.8.0
> torch               ✅ 2.7.1!=2.4.0,>=1.8.0; sys_platform == "win32"
> torchvision         ✅ 0.22.1>=0.9.0
> tqdm                ✅ 4.67.1>=4.64.0
> psutil              ✅ 7.0.0
> py-cpuinfo          ✅ 9.0.0
> pandas              ✅ 2.3.0>=1.1.4
> ultralytics-thop    ✅ 2.0.14>=2.0.0
> 
> 
> ### Laptop:
> 
> Ultralytics 8.3.162 🚀 Python-3.12.11 torch-2.7.1+cu126 CUDA:0 (NVIDIA GeForce GTX 850M, 4038MiB)
> Setup complete ✅ (8 CPUs, 15.5 GB RAM, 435.6/455.0 GB disk)
> 
> OS                  Linux-6.8.0-60-generic-x86_64-with-glibc2.35
> Environment         Linux
> Python              3.12.11
> Install             git
> Path                /home/developer/zekri/projects/gitlab/vms/raspberry/ai/anpr/.venv/lib/python3.12/site-packages/ultralytics
> RAM                 15.51 GB
> Disk                435.6/455.0 GB
> CPU                 Intel Core(TM) i7-4710HQ 2.50GHz
> CPU count           8
> GPU                 NVIDIA GeForce GTX 850M, 4038MiB
> GPU count           1
> CUDA                12.6
> 
> numpy               ✅ 2.2.6>=1.23.0
> matplotlib          ✅ 3.10.3>=3.3.0
> opencv-python       ✅ 4.11.0.86>=4.6.0
> pillow              ✅ 11.2.1>=7.1.2
> pyyaml              ✅ 6.0.2>=5.3.1
> requests            ✅ 2.32.4>=2.23.0
> scipy               ✅ 1.16.0>=1.4.1
> torch               ✅ 2.7.1>=1.8.0
> torch               ✅ 2.7.1!=2.4.0,>=1.8.0; sys_platform == "win32"
> torchvision         ✅ 0.22.1>=0.9.0
> tqdm                ✅ 4.67.1>=4.64.0
> psutil              ✅ 7.0.0
> py-cpuinfo          ✅ 9.0.0
> pandas              ✅ 2.3.0>=1.1.4
> ultralytics-thop    ✅ 2.0.14>=2.0.0
> 

haye you try to enable stream=True when you doing the inference? 

results = model(...., stream=True)

**Comment by Y-T-G** [comment]
I think it's a OpenVINO issue. You can try opening an issue there.

**Comment by erfan-zekri** [comment]
> I think it's a OpenVINO issue. You can try opening an issue there.

You're right!
Here is the link of [Issue on OpenVINO](https://github.com/openvinotoolkit/openvino/issues/31064)

**Comment by glenn-jocher** [comment]
Thanks for opening the [OpenVINO issue](https://github.com/openvinotoolkit/openvino/issues/31064) - that's the right approach since this appears to be platform-specific behavior with OpenVINO on ARM64. While waiting for their response, you could try using `stream=True` in your inference calls as @alien087 suggested, which creates a generator that's more memory-efficient for continuous processing: `results = model(source, stream=True)`.

**Comment by abhiagwl4262** [comment]
Its not only with openVINO. The memory leak is there even for pytorch training. I often faced process killed because of memory issue after few epochs of training.

**Comment by glenn-jocher** [comment]
@abhiagwl4262 thanks for the report—this is unrelated to the OpenVINO inference thread; please open a new issue for PyTorch training memory growth and include your exact `yolo train ...` command or Python snippet, full text logs/traceback, the outputs below, and whether you’re training in a notebook (if so, restart the kernel), plus any quick checks you tried like reducing `batch` and setting `workers=0`; for reference, see the [Train mode docs](https://docs.ultralytics.com/modes/train/) and the [common issues guide](https://docs.ultralytics.com/guides/yolo-common-issues/).

```bash
yolo checks
```

```python
from ultralytics.utils.checks import collect_system_info
print(collect_system_info())
```
